File: MyJournal/app/src/main/java/com/smart/journal/module/write/db/JournalDBHelper.java
package com.smart.journal.module.write.db;

import android.text.TextUtils;

import com.smart.journal.contants.Contancts;
import com.smart.journal.db.AppDatabase;
import com.smart.journal.db.dao.JournalDao;
import com.smart.journal.module.write.bean.JournalBean;
import com.smart.journal.module.write.bean.JournalBeanDBBean;
import com.smart.journal.tools.location.LocationTools;
import com.smart.journal.tools.location.bean.LocationBean;

import java.util.Date;
import java.util.List;
import java.util.UUID;

import io.realm.Realm;
import io.realm.RealmResults;
import io.realm.Sort;

/**
 * @author guandongchen
 * @date 2018/1/22
 */

public class JournalDBHelper {

    /**
     * 保存日记
     * @param writeSectionBeans
     */
    public static void saveJournal(List<JournalBean> writeSectionBeans ){
        JournalBeanDBBean journalBeanDBBean = new JournalBeanDBBean();
        StringBuilder contentSb = new StringBuilder();
        for (JournalBean journalBean:
            writeSectionBeans) {

            if (journalBean.getItemType()==JournalBean.WRITE_TAG_IMAGE){

                contentSb.append(Contancts.FILE_TYPE_IMAGE+journalBean.getImageURL()+Contancts.FILE_TYPE_SPLIT);
            }else{
                if (!TextUtils.isEmpty(journalBean.getContent().trim())){
                    contentSb.append(Contancts.FILE_TYPE_TEXT+journalBean.getContent()+Contancts.FILE_TYPE_SPLIT);
                }

            }
        }
        journalBeanDBBean.setContent(contentSb.toString());
        journalBeanDBBean.setDate(System.currentTimeMillis());
        LocationBean locationBean =  LocationTools.getLocationBean();
        if (!TextUtils.isEmpty(locationBean.getAdress())){
            journalBeanDBBean.setAddress(locationBean.getAdress());
            journalBeanDBBean.setLatitude(locationBean.getLatitude());
            journalBeanDBBean.setLongitude(locationBean.getLongitude());
        }
        journalBeanDBBean.setTags("默认");
        AppDatabase.Companion.getInstance().mJournalDao().saveJournal(journalBeanDBBean);
    }


    /**
     * 获取所有日记
     * @return
     */
    public static List<JournalBeanDBBean> getAllJournals(JournalBeanDBBean journalBeanDBBean){
        return AppDatabase.Companion.getInstance().mJournalDao().getAllJournal();
    }

    /**
     * 删除日记
     * @param realm
     * @param journalBeanDBBean
     */
    public static void deleteJournal(Realm realm,JournalBeanDBBean journalBeanDBBean){

        realm.executeTransaction(realm1 -> journalBeanDBBean.deleteFromRealm());
    }
}

File: MyJournal/app/src/main/java/com/smart/journal/module/write/db/JournalDBHelper.kt
package com.smart.journal.module.write.db

import android.text.TextUtils
import androidx.lifecycle.LiveData
import com.smart.journal.app.MyApp
import com.smart.journal.contants.Contancts
import com.smart.journal.db.dao.JournalDao
import com.smart.journal.db.entity.JournalBeanDBBean
import com.smart.journal.module.write.bean.JournalBean
import com.smart.journal.module.write.bean.WriteSettingBean
import com.smart.journal.tools.location.LocationTools


/**
 * @author guandongchen
 * @date 2018/1/22
 */

object JournalDBHelper {

    private var journalDao: JournalDao? = MyApp.database!!.mJournalDao()

    /**
     * 获取所有日记
     * @return
     */
    fun allJournals():LiveData<List<JournalBeanDBBean>>{
       return journalDao!!.allJournal()
    }

    /**
     * 保存日记
     * @param writeSectionBeans
     */
    fun saveJournal(writeSectionBeans: List<JournalBean>, settingBean: WriteSettingBean?) {
        val journalBeanDBBean = JournalBeanDBBean()
        val contentSb = StringBuilder()
        for (journalBean in writeSectionBeans) {

            if (journalBean.itemType == JournalBean.WRITE_TAG_IMAGE) {

                contentSb.append(Contancts.FILE_TYPE_IMAGE + journalBean.imageURL + Contancts.FILE_TYPE_SPLIT)
            } else {
                contentSb.append(Contancts.FILE_TYPE_TEXT + journalBean.content + Contancts.FILE_TYPE_SPLIT)

            }
        }
        journalBeanDBBean.content = contentSb.toString()
        journalBeanDBBean.date = settingBean!!.time!!
        if (settingBean.journalBook != null) {
            journalBeanDBBean.bookId = settingBean!!.journalBook!!.id
        }
        if (settingBean!!.location != null) {
            val locationBean = settingBean!!.location
            if (!TextUtils.isEmpty(locationBean!!.title)) {
                journalBeanDBBean.address = locationBean.title
                journalBeanDBBean.latitude = locationBean.latitude
                journalBeanDBBean.longitude = locationBean.longitude
            }
        } else {
            val locationBean = LocationTools.locationBean
            if (!TextUtils.isEmpty(locationBean!!.adress)) {
                journalBeanDBBean.address = locationBean.adress
                journalBeanDBBean.latitude = locationBean.latitude
                journalBeanDBBean.longitude = locationBean.longitude
            }
        }

        settingBean.tags?.let {
            journalBeanDBBean.tags = it.toString().replace("[", "").replace("]", "")
        }
        settingBean.journalId?.let {
            journalBeanDBBean.id = it
        }
        journalBeanDBBean.favourite = settingBean.isFavourite
        journalDao!!.saveJournal(journalBeanDBBean)
    }


    /**
     * 删除日记
     * @param journalBeanDBBean
     */
    fun deleteJournal(journalBeanDBBean: JournalBeanDBBean) {

        journalDao!!.deleteJournal(journalBeanDBBean)
    }

    fun deleteJournalById(journalId: Int) {
        journalDao!!.deleteJournalById(journalId)
    }

    fun queryJournalByDate(date: Long): List<JournalBeanDBBean> {
        return journalDao!!.queryJournalByDate(date)
    }

    fun queryJournalById(id: Int): List<JournalBeanDBBean> {
        return journalDao!!.queryJournalById(id)
    }

    fun searchJournalByKeyWord(keyWord: String): LiveData<List<JournalBeanDBBean>> {
        return journalDao!!.searchJournalByKeyWord(keyWord)
    }

    fun searchJournalByTag(tagName: String): LiveData<List<JournalBeanDBBean>> {
        return journalDao!!.searchJournalByTag(tagName)
    }

    fun searchJournalByLocationName(locatonName: String): LiveData<List<JournalBeanDBBean>> {
        return journalDao!!.searchJournalByTagLocationName(locatonName)
    }

    fun searchJournalFromFavourite(keyWord: String): LiveData<List<JournalBeanDBBean>> {
        return journalDao!!.searchJournalFromFavourite(keyWord)
    }

}

